opennms-dao/src/main/java/org/opennms/netmgt/dao/hibernate/OutageDaoHibernate.java

```diff
@@ -103,6 +115,66 @@ public OnmsOutage currentOutageForService(OnmsMonitoredService service) {
         });
     }
 
+    private Stream<OnmsOutage> initializeOutage(Stream<OnmsOutage> outages) {
+        return outages
+                .map(outage -> {
+                    Hibernate.initialize(outage.getMonitoredService());
+                    Hibernate.initialize(outage.getMonitoredService().getIpInterface());
+                    Hibernate.initialize(outage.getMonitoredService().getIpInterface().getNode());
+                    Hibernate.initialize(outage.getMonitoredService().getServiceType());
+                    Hibernate.initialize(outage.getForeignId());
+                    return outage;
+                })
+                .sorted(OUTAGE_COMPARATOR);
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public Collection<OnmsOutage> matchingOutages(final ServiceSelector selector) {
+        final Set<InetAddress> matchingAddrs = new HashSet<InetAddress>(m_filterDao.getIPAddressList(selector.getFilterRule()));
+        @SuppressWarnings("unchecked")
+        final Collection<OnmsOutage> outages = (Collection<OnmsOutage>)getHibernateTemplate().findByNamedParam("FROM OnmsOutage AS o WHERE o.monitoredService.serviceType.name IN (:services)","services", selector.getServiceNames());
+        return initializeOutage(
+                                outages.parallelStream()
+                                .filter(outage -> matchingAddrs.contains(outage.getMonitoredService().getIpAddress()))
+                ).collect(Collectors.toList());
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public Collection<OnmsOutage> matchingLatestOutages(final ServiceSelector selector) {
+        final Set<String> matchingAddrs = m_filterDao.getIPAddressList(selector.getFilterRule()).stream().map(InetAddressUtils::str).collect(Collectors.toSet());
+        final List<Integer> outageIds = getHibernateTemplate().execute(new HibernateCallback<List<Integer>>() {
+            @Override
+            @SuppressWarnings("unchecked")
+            public List<Integer> doInHibernate(Session session) throws HibernateException, SQLException {
+                return (List<Integer>) session.createSQLQuery(
+                    "SELECT DISTINCT MAX(outageID) AS outageID\n" +
```

Comment:
It seems inefficient to have to pull all of the latest outage ids for all of the services from the database when were only actually interested in a small subset of these.

Reply:
I'll see what I can do.  Every time I tried to get fancy with this query it started doing the weird insert thing, but that's when I was doing it in HQL.